### DISCUSSION TITLE
remove and check container

### DISCUSSION THREAD

**Discussion by mauroForlimpopoli** [Moddable-OpenSource/moddable#725] [Q&A]
Good morning,
First my container was in my main.js. Then I move it in my assets.js
I add it in my main.is with
application.add(new ASSETS.SwitchScreen(value));

1) How can I remove it ?
PS: I can't use something like:
application.remove(application.first);
but I need something like:
application.remove(SwitchScreen);

2) How can I check if is it visible?
normally I used:
SwitchScreen.index
result = -1 if not visible
but now it doesn't work

**Comment by lprader** [comment]
> 1. How can I remove it ?
> PS: I can't use something like:
> application.remove(application.first);
> but I need something like:
> application.remove(SwitchScreen);

You just need a reference to the `SwitchScreen` object. For example, you could do this:

```js
let myScreen = new ASSETS.SwitchScreen(value);
application.add(myScreen);
application.remove(myScreen);
```

But there are other ways to get references to content objects in a container you might prefer too. I wrote a summary in [this comment](https://github.com/Moddable-OpenSource/moddable/discussions/721#discussioncomment-1444687) of methods described in the  Piu chapter of the [book](https://www.moddable.com/book) @phoddie and I wrote. It has a 6 page long section called "Accessing Content Objects in a Container". If you don't want to use `application.first`, `application.first.next`, etc. you might try using the `content` method of a container (see the [Piu documentation](https://github.com/Moddable-OpenSource/moddable/blob/public/documentation/piu/piu.md) for a full description of what the `content` method does). For example, you could do this:

```js
application.add(new ASSETS.SwitchScreen(value, {name: "myScreen"}));
let myScreen = application.content("myScreen");
application.remove(myScreen);
```

Here's what's happening in the code above:
- The first line passes in a `name` property to the `SwitchScreen` constructor. This sets the `name` property of the created instance to `"myScreen"`.
- The second line gets a reference to the `SwitchScreen` instance using the application's `content` method
- The third line removes the screen from the application

> 2. How can I check if is it visible?
> normally I used:
> SwitchScreen.index
> result = -1 if not visible
> but now it doesn't work

It depends what you mean by visible ;)

1. A content object has to be part of the "containment hierarchy" to appear on the display. If you want to know whether a content object has been added to the containment hierarchy, you can just check its `container` property. If it's `undefined`, it is not in a container and is therefore not in the containment hierarchy. If its `container` property is an object, it is in the containment hierarchy. For example:

```js
let myScreen = new ASSETS.SwitchScreen(value);
let foo = myScreen.container;    // foo is undefined because `myScreen` hasn't been added to the application

application.add(myScreen);
foo = myScreen.container;    // foo is now a reference to the `application` object

application.remove(myScreen);
foo = myScreen.container;    // foo is undefined again because it has been removed from the application
```

2. Content objects have a `visible` property. This property isn't used very often, but you can use it if you want a content object to be in the containment hierarchy, but you don't want the content object to actually be drawn on the display. The `visible` property defaults to `true`, but you can simply set it to `false` to prevent the content from appearing on the display. For example:

```js
let myScreen = new ASSETS.SwitchScreen(value);
application.add(myScreen);
myScreen.visible = false;
```

3. A content object will also be "invisible" if its coordinates are beyond the dimensions of the display. For example, if you're running an application on a 320x240 screen, `myScreen` in the example below won't appear because its `left` property puts it beyond the dimensions of the screen:

```js
let myScreen = new ASSETS.SwitchScreen(value, {left: 320});
application.add(myScreen);
```

Checking whether a content object is within the bounds of the display might look something like this:

```js
let isPartiallyVisible = myScreen.x <= application.width;
let isFullyVisible = (myScreen.x + myScreen.width) <= application.width;
```

**Reply by mauroForlimpopoli** [reply]
Thank you Lizzie for your esaustive answers. 
How can I trace the value of a Object
trace(myScreen.container) is [object Object]

**Reply by lprader** [reply]
What do you mean by "the value of an Object"? What are you hoping to see from that `trace` statement?

**Reply by mauroForlimpopoli** [reply]
I mean something like:
alert (JSON.stringify(myScreen, null, 4));  // in javascript

Refers to your indication:
```
let myScreen = new SwitchScreen(value);
application.add(myScreen);
```
and then
```
application.remove(myScreen);
```
SwitchScreen is imported form ASSETS

it works well, but if the instruction in nested in a function inside a function it seams doesn't work, but no error is showed.
I will check again if I did some errors
Thank you for your assistance
